exec.go

```diff
@@ -91,7 +108,118 @@ EXAMPLE:
 		},
 	},
 	Action: func(context *cli.Context) error {
-		// TODO
-		return nil
+		params, err := validExecParams(context)
+		if err != nil {
+			return err
+		}
+
+		return execute(params)
 	},
 }
+
+func validExecParams(context *cli.Context) (execParams, error) {
+	ctxArgs := context.Args()
+
+	params := execParams{
+		cID:          ctxArgs.First(),
+		pidFile:      context.String("pid-file"),
+		console:      context.String("console"),
+		detach:       context.Bool("detach"),
+		processLabel: context.String("process-label"),
+		noSubreaper:  context.Bool("no-subreaper"),
+	}
+
+	if context.IsSet("process") == true {
+		var ociProcess specs.Process
+
+		fileContent, err := ioutil.ReadFile(context.String("process"))
+		if err != nil {
+			return execParams{}, err
+		}
+
+		if err := json.Unmarshal(fileContent, ociProcess); err != nil {
+			return execParams{}, err
+		}
+
+		params.ociProcess = ociProcess
+	} else {
+		params.ociProcess = specs.Process{
+			Terminal: context.Bool("tty"),
+			User: specs.User{
+				Username: context.String("user"),
+			},
+			Args:            ctxArgs.Tail(),
+			Env:             context.StringSlice("env"),
+			Cwd:             context.String("cwd"),
+			NoNewPrivileges: context.Bool("no-new-privs"),
+			ApparmorProfile: context.String("apparmor"),
+		}
+	}
+
+	return params, nil
+}
+
+func execute(params execParams) error {
+	if err := validContainer(params.cID); err != nil {
+		return err
+	}
+
+	podStatus, err := vc.StatusPod(params.cID)
+	if err != nil {
+		return err
+	}
+
+	if len(podStatus.ContainersStatus) != 1 {
+		return fmt.Errorf("BUG: ContainerStatus list from PodStatus %s is wrong, expecting only one ContainerStatus: %v", params.cID, podStatus.ContainersStatus)
+	}
+
+	// container MUST be running
+	if podStatus.ContainersStatus[0].State.State != vc.StateRunning {
+		return fmt.Errorf("Container %s is not running", params.cID)
+	}
+
+	// Check status of process running inside the container
+	running, err := processRunning(podStatus.ContainersStatus[0].PID)
+	if err != nil {
+		return err
+	}
+	if running == false {
+		if err := stopContainer(podStatus); err != nil {
+			return err
+		}
+
+		return fmt.Errorf("Process not running inside container %s", params.cID)
+	}
+
+	envVars, err := oci.EnvVars(params.ociProcess.Env)
+	if err != nil {
+		return err
+	}
+
+	cmd := vc.Cmd{
+		Args:    params.ociProcess.Args,
+		Envs:    envVars,
+		WorkDir: params.ociProcess.Cwd,
+		User:    params.ociProcess.User.Username,
+	}
+
+	_, process, err := vc.EnterContainer(params.cID, podStatus.ContainersStatus[0].ID, cmd)
+	if err != nil {
+		return err
+	}
+
+	// Start the shim to retrieve its PID.
+	pid, err := startShim(process)
+	if err != nil {
+		return err
+	}
+
+	// Creation of PID file has to be the last thing done in the exec
+	// because containerd considers the exec complete after this file
```

Comment:
Well, `containerd` considers the `exec` to have "finished starting", not "finished".